File: server/src/main/java/de/ketrwu/portfolio/controller/content/TextPageController.java
package de.ketrwu.portfolio.controller.content;

import lombok.extern.slf4j.Slf4j;
import org.springframework.core.io.Resource;
import org.springframework.core.io.support.PathMatchingResourcePatternResolver;
import org.springframework.core.io.support.ResourcePatternResolver;
import org.springframework.stereotype.Controller;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.PathVariable;
import org.springframework.web.bind.annotation.RequestMapping;

import java.io.IOException;

@Slf4j
@Controller
@RequestMapping
public class TextPageController {

    @GetMapping("/page/{page}")
    public String getTextPage(@PathVariable String page) {
        ClassLoader cl = this.getClass().getClassLoader();
        ResourcePatternResolver resolver = new PathMatchingResourcePatternResolver(cl);
        try {
            for (Resource resource : resolver.getResources("classpath*:/templates/content/text/*.html")) {
                if (resource.getFilename().replace(".html", "").equalsIgnoreCase(page)) {
                    return "content/text/" + page.toLowerCase();
                }
            }
        } catch (IOException e) {
            log.error("Failed to load text-pages from classpath", e);
        }
        return "content/text/error";
    }

}

File: server/src/main/kotlin/de/ketrwu/portfolio/controller/content/TextPageController.kt
package de.ketrwu.portfolio.controller.content

import de.ketrwu.portfolio.service.TextPageService
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.stereotype.Controller
import org.springframework.web.bind.annotation.GetMapping
import org.springframework.web.bind.annotation.PathVariable
import org.springframework.web.bind.annotation.RequestMapping

/**
 * Controller to serve pages in templates/contet/text/ dynamically
 * @author Kenneth Wußmann
 */
@Controller
@RequestMapping
class TextPageController {

    @Autowired
    private val textPageService: TextPageService? = null

    /**
     * Mapping that looks for path-variable page in classpath and serves it
     */
    @GetMapping("/page/{page}")
    fun getTextPage(@PathVariable page: String, model: MutableMap<String, Any>): String {
        return when {
            textPageService!!.isHtmlTextPage(page) -> "content/text/" + page.toLowerCase()
            textPageService.isMarkdownTextPage(page) -> {
                model["page"] = textPageService.renderMarkdown(page)
                "content/text/markdown"
            }
            else -> "content/text/error"
        }
    }

    /**
     * Mapping that looks for path-variable page in classpath and serves it at its tags-based path
     */
    @GetMapping("/page/{tag}/{page}")
    fun getTaggedTextPage(
        @PathVariable tag: String,
        @PathVariable page: String,
        model: MutableMap<String, Any>
    ): String {
        return when {
            textPageService!!.isMarkdownTextPage(page, tag) -> {
                model["page"] = textPageService.renderMarkdown(page)
                "content/text/markdown"
            }
            else -> "content/text/error"
        }
    }
}
